### ISSUE TITLE
Feature: Remote Execution Environment Backends (SSH, Modal, Daytona)

### ISSUE THREAD

**Issue by MervinPraison** [MervinPraison/PraisonAI#1395]
## Overview

Add remote execution environment backends to enable agents to run code on remote machines, cloud functions, and container platforms beyond local Docker.

**Current state:** PraisonAI has `SandboxProtocol` in core SDK and `DockerSandbox` implementation in wrapper.

**Gap:** Missing SSH, Modal, Daytona, and Singularity backends for remote/cloud code execution.

---

### Why This Matters
- **SSH Backend**: Execute code on remote servers (dev boxes, GPU machines, production servers)
- **Modal Backend**: Serverless GPU execution for ML workloads
- **Daytona Backend**: Cloud development environments
- **Singularity Backend**: HPC cluster execution (scientific computing)

### Current Implementation

**Core SDK** (`praisonaiagents/sandbox/`):
- `protocols.py` — `SandboxProtocol` (360 lines)
- `config.py` — `SandboxConfig`, `ResourceLimits`

**Wrapper** (`praisonai/sandbox/`):
- `docker.py` — `DockerSandbox` implementation
- `subprocess.py` — Local subprocess sandbox
- `sandlock.py` — Sandlock integration

---

### Existing Protocol

```python
@runtime_checkable
class SandboxProtocol(Protocol):
    @property
    def is_available(self) -> bool: ...
    @property
    def sandbox_type(self) -> str: ...
    
    async def start(self) -> None: ...
    async def stop(self) -> None: ...
    async def execute(self, code: str, language: str = "python", 
                      limits: Optional[ResourceLimits] = None, ...) -> SandboxResult: ...
    async def run_command(self, command: Union[str, List[str]], ...) -> SandboxResult: ...
    async def write_file(self, path: str, content: Union[str, bytes]) -> bool: ...
    async def read_file(self, path: str) -> Optional[Union[str, bytes]]: ...
    async def cleanup(self) -> None: ...
```

### Key Design Decisions

1. **Protocol stays in core** — No changes to `praisonaiagents/sandbox/protocols.py`
2. **Implementations in wrapper** — All backends go in `praisonai/sandbox/`
3. **Lazy imports** — Each backend only imports its deps when instantiated
4. **Optional dependencies** — Add `[ssh]`, `[modal]`, `[daytona]` extras

---

### Phase 1: SSH Backend (High Priority)

```python
# praisonai/sandbox/ssh.py
class SSHSandbox:
    """Execute code on remote servers via SSH.
    
    Example:
        sandbox = SSHSandbox(
            host="gpu-server.example.com",
            user="ubuntu",
            key_file="~/.ssh/id_rsa",
        )
        result = await sandbox.execute("python train.py")
    """
    
    def __init__(
        self,
        host: str,
        user: str = "root",
        port: int = 22,
        key_file: Optional[str] = None,
        password: Optional[str] = None,
        working_dir: str = "/tmp/praisonai",
    ): ...
```

**Dependencies:** `paramiko` or `asyncssh`

### Phase 2: Modal Backend (Medium Priority)

```python
# praisonai/sandbox/modal.py
class ModalSandbox:
    """Execute code on Modal serverless GPUs.
    
    Example:
        sandbox = ModalSandbox(gpu="A100")
        result = await sandbox.execute("import torch; print(torch.cuda.is_available())")
    """
    
    def __init__(
        self,
        gpu: Optional[str] = None,  # "T4", "A10G", "A100"
        image: str = "python:3.11",
        timeout: int = 300,
    ): ...
```

**Dependencies:** `modal`

### Phase 3: Daytona Backend (Lower Priority)

```python
# praisonai/sandbox/daytona.py
class DaytonaSandbox:
    """Execute code in Daytona cloud dev environments."""
```

**Dependencies:** `daytona`

---

### New Files

| File | Purpose |
|------|---------|
| `praisonai/sandbox/ssh.py` | SSH remote execution |
| `praisonai/sandbox/modal.py` | Modal serverless GPU |
| `praisonai/sandbox/daytona.py` | Daytona cloud dev |
| `tests/unit/sandbox/test_ssh.py` | SSH backend tests |
| `tests/unit/sandbox/test_modal.py` | Modal backend tests |

### Modified Files

| File | Change |
|------|--------|
| `praisonai/sandbox/__init__.py` | Add lazy exports for new backends |
| `pyproject.toml` | Add optional deps: `[ssh]`, `[modal]`, `[daytona]` |
| `docs/features/sandbox.mdx` | Document new backends |

---

### Dependencies

```toml
[project.optional-dependencies]
ssh = ["asyncssh>=2.14.0"]
modal = ["modal>=0.64.0"]
daytona = ["daytona>=0.148.0"]
```

### Performance Impact
- **Zero** when backends not used (lazy imports)
- SSH: ~50ms connection overhead per command
- Modal: ~2-5s cold start, <100ms warm

### Safety / Approval
- SSH: Requires explicit host/credentials — no auto-discovery
- Modal: Uses Modal's built-in sandboxing
- All backends respect `ResourceLimits` from protocol

### Multi-Agent Safety
- Each agent gets isolated sandbox instance
- No shared state between sandbox instances
- Connection pooling optional (SSH only)

---

## Acceptance Criteria

- [ ] `SSHSandbox` implements `SandboxProtocol`
- [ ] `ModalSandbox` implements `SandboxProtocol`
- [ ] `DaytonaSandbox` implements `SandboxProtocol`
- [ ] All backends have unit tests (mocked)
- [ ] Integration tests for SSH (requires test server)
- [ ] Documentation in `docs/features/sandbox.mdx`
- [ ] Examples in `examples/python/sandbox_remote.py`
- [ ] Optional deps in `pyproject.toml`
- [ ] Lazy imports — no import-time overhead

---

### Key Files to Read First

1. `praisonaiagents/sandbox/protocols.py` (360 lines) — Protocol definition
2. `praisonai/sandbox/docker.py` (~200 lines) — Reference implementation
3. `praisonai/sandbox/__init__.py` — Export pattern

### Testing Commands

```bash
# Unit tests (mocked)
pytest tests/unit/sandbox/test_ssh.py -v

# Integration test (requires SSH server)
SSH_TEST_HOST=localhost pytest tests/integration/sandbox/test_ssh_integration.py -v

# Smoke test
python -c "from praisonai.sandbox import SSHSandbox; print('OK')"
```

---

## Priority

**Medium** — Docker covers most use cases. SSH is the highest priority for remote GPU access.

## Labels

`enhancement`, `sandbox`, `remote-execution`
